### PR TITLE
[REVIEW] Remove legacy libcudf

### LINKED ISSUE #3031: [FEA] Remove unused all_bitmask_on function
**Is your feature request related to a problem? Please describe.**

The function all_bitmask_on, defined in cpp/src/bitmask/legacy/bitmask_ops.cu, appears to be unused/dead code.

**Describe the solution you'd like**

This function could be removed.

**Describe alternatives you've considered**

The containing file is in a legacy folder, so the entire file may be deleted at some point.

### LINKED ISSUE #3045: [BUG] Remove all `cudaDeviceSynchronize` from libcudf features
**Is your feature request related to a problem? Please describe.**

`cudaDeviceSynchronize()`s are littered throughout libcudf code. These are a heavy hammer that obliterate any opportunity for multi-threaded/multi-stream use of libcudf.

**Describe the solution you'd like**
Any `cudaDeviceSynrhonize()` should be removed in favor of a more appropriate synchronization, e.g., `cudaStreamSynchronize()`.

### LINKED ISSUE #4050: [FEA] Remove legacy libcudf code
**Is your feature request related to a problem? Please describe.**

The legacy libcudf code should no longer be used and therefore removed. 

**Describe the solution you'd like**

Once the porting to use libcudf++ APIs is complete, the legacy code should be removed.

### PR DESCRIPTION
**Pull request by devavret**
Depends on #4627.
Closes #4050
Closes #3031
Closes #3045

### CODE REVIEW

**Comment by trevorsm7** [comment]
@devavret it looks like you may be able close https://github.com/rapidsai/cudf/issues/3045 with this PR as well

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by devavret** [comment]
How did all these extra commits appear here?

**Comment by harrism** [comment]
> How did all these extra commits appear here?

I told you I pushed directly to your PR.

**Comment by devavret** [comment]
> > How did all these extra commits appear here?
> 
> I told you I pushed directly to your PR.

No I meant the commits from Java. I never merged any Java PR into this. Anyway, merging from branch 0.14 fixed it. But the request to review from java code owners still went out.

**Comment by vuule** [comment]
From the failing build log:
`/cpp/benchmarks/join/legacy/join_benchmark.cu:18:23: fatal error: cudf/cudf.h: No such file or directory`
Did join_benchmark maybe slip through the cracks?

**Comment by devavret** [comment]
> From the failing build log:
> `/cpp/benchmarks/join/legacy/join_benchmark.cu:18:23: fatal error: cudf/cudf.h: No such file or directory`
> Did join_benchmark maybe slip through the cracks?

It's removed from my workspace. Investigating.

**Comment by harrism** [comment]
I'm working on it.

File: cpp/tests/replace/replace_tests.cu [hunk 1]
```diff
@@ -1,4 +1,6 @@
 /*
+ * Copyright 2019, NVIDIA CORPORATION.
```
Comment (trevorsm7) [inline cpp/tests/replace/replace_tests.cu:2]:
```suggestion
 * Copyright 2019-20, NVIDIA CORPORATION.
```

File: cpp/tests/replace/replace_tests.cu [hunk 2]
```diff
@@ -299,7 +295,7 @@ struct ReplaceTest : cudf::test::BaseFixture {
   ~ReplaceTest() {}
 };
 
-/* --------------------------------------------------------------------------*
+/** --------------------------------------------------------------------------*
```
Comment (trevorsm7) [inline cpp/tests/replace/replace_tests.cu:298]:
```suggestion
/**
```

File: cpp/tests/utilities/cudf_gtest.hpp [hunk 3]
```diff
@@ -116,3 +119,17 @@ struct TypeList<Types<TYPES...>> {
 
 #define CUDA_EXPECT_THROW_MESSAGE(x, msg) \
   EXPECT_THROW_MESSAGE(x, cudf::cuda_error, "CUDA error encountered at:", msg)
+
+/**---------------------------------------------------------------------------*
```
Comment (trevorsm7) [inline cpp/tests/utilities/cudf_gtest.hpp:123]:
```suggestion
/**
```

**Comment by trevorsm7** [comment]
Looks like LEGACY_JOIN_BENCH and LEGACY_SEARCH_BENCH_SRC still need to be removed from benchmarks/CMakeLists.txt

**Comment by vuule** [comment]
@devavret Can you please also remove `RMM_TRY` and `throw_rmm_error` from error.hpp? 
AFAICT they are only used in legacy code, and use a deprecated rmm function (causing the warning avalanche).

**Comment by harrism** [comment]
I will do it.

**Pull request merged**